cluster-operator/src/test/java/io/strimzi/operator/cluster/operator/resource/ZooKeeperRollerTest.java

```diff
@@ -78,6 +80,63 @@ public void testAllPodsAreRolled(VertxTestContext context)  {
+        when(podOperator.isReady(any(), eq("my-cluster-zookeeper-0"))).thenReturn(true);
+        when(podOperator.isReady(any(), eq("my-cluster-zookeeper-1"))).thenReturn(false);
+        when(podOperator.isReady(any(), eq("my-cluster-zookeeper-2"))).thenReturn(true);
+        when(podOperator.listAsync(any(), any(Labels.class))).thenReturn(Future.succeededFuture(PODS));
+
+        ZookeeperLeaderFinder leaderFinder = mock(ZookeeperLeaderFinder.class);
+        when(leaderFinder.findZookeeperLeader(any(), any(), any(), any())).thenReturn(Future.succeededFuture("my-cluster-zookeeper-2"));
+
+        MockZooKeeperRoller roller = new MockZooKeeperRoller(podOperator, leaderFinder, 300_00L);
+
+        Checkpoint async = context.checkpoint();
+        roller.maybeRollingUpdate(Reconciliation.DUMMY_RECONCILIATION, DUMMY_SELECTOR, shouldRoll, new Secret(), new Secret())
+              .onComplete(context.succeeding(v -> context.verify(() -> {
+                  assertThat(roller.podRestarts.size(), is(3));
+                  assertThat(roller.podRestarts.pollFirst(), is("my-cluster-zookeeper-1"));
+                  assertThat(roller.podRestarts.pollFirst(), is("my-cluster-zookeeper-0"));
+                  assertThat(roller.podRestarts.pollFirst(), is("my-cluster-zookeeper-2"));
+                  async.flag();
+              })));
+    }
+
+    @Test
+    public void testReadinessOfNonRestartedPodCanPreventFurtherRestarts(VertxTestContext context)  {
+        Function<Pod, List<String>> shouldRoll = pod -> {
+            if (!"my-cluster-zookeeper-1".equals(pod.getMetadata().getName()))  {
+                return List.of("Should restart");
+            } else {
+                return List.of();
+            }
+        };
+        PodOperator podOperator = mock(PodOperator.class);
+        when(podOperator.isReady(any(), eq("my-cluster-zookeeper-0"))).thenReturn(true);
+        when(podOperator.isReady(any(), eq("my-cluster-zookeeper-1"))).thenReturn(false);
+        when(podOperator.isReady(any(), eq("my-cluster-zookeeper-2"))).thenReturn(true);
+        when(podOperator.listAsync(any(), any(Labels.class))).thenReturn(Future.succeededFuture(PODS));
+        when(podOperator.readiness(any(), any(), any(), anyLong(), anyLong())).thenReturn(Future.failedFuture("failure"));
+
+
+        ZookeeperLeaderFinder leaderFinder = mock(ZookeeperLeaderFinder.class);
+        when(leaderFinder.findZookeeperLeader(any(), any(), any(), any())).thenReturn(Future.succeededFuture("my-cluster-zookeeper-2"));
+
+        MockZooKeeperRoller roller = new MockZooKeeperRoller(podOperator, leaderFinder, 300_00L);
+
+        Checkpoint async = context.checkpoint();
+        roller.maybeRollingUpdate(Reconciliation.DUMMY_RECONCILIATION, DUMMY_SELECTOR, shouldRoll, new Secret(), new Secret())
+              .onComplete(context.failing(v -> context.verify(() -> {
+                  assertThat(roller.podRestarts.size(), is(0));
```

Comment:
Size 0 is a bit of a surprise, I was expecting `my-cluster-zookeeper-0` to be restartable. Is it zero because `my-cluster-zookeeper-1` is false and thus rolling any nodes would break the cluster quorum?

Reply:
Correct, `my-cluster-zookeeper-1` is non-ready, so it is sorted to the front when we are iterating over the pods. We then await readiness on it, which is mocked to fail. This failure stops the chain and nothing is restarted.

Previously it calculated that the non-ready pods had no restart reasons so they were skipped over and the next eligible pod would be restarted. This can eventually take 100% of the ZK pods offline.

Reply:
I'll also note down that maybe there is another hole/race, if the leader is somehow non-ready we could restart one of the non-leader nodes because we always restart the leader last.

We could change the initial loop to also await readiness of the leader, or put it off to another PR since it's an existing problem.

Reply:
I changed it to also await readiness of the leader when doing the first iteration of the pods.